packages/web-components/fast-element/src/context.ts

```diff
@@ -0,0 +1,272 @@
+import type { Constructable } from "./interfaces.js";
+import { Metadata } from "./metadata.js";
+
+/**
+ * A Context object defines an optional initial value for a Context, as well as a name identifier for debugging purposes.
+ * @public
+ */
+export type Context<T> = {
+    readonly name: string;
+    readonly initialValue?: T;
+};
+
+/**
+ * A constant key that can be used to represent a Context dependency.
+ * The key can be used for context or DI but also doubles as a decorator for
+ * resolving the associated dependency.
+ * @beta
+ */
+export type ContextDecorator<T = any> = Readonly<Context<T>> &
+    PropertyDecorator &
+    ParameterDecorator;
+
+/**
+ * A Context object defines an optional initial value for a Context, as well as a name identifier for debugging purposes.
+ * The FASTContext can also be used as a decorator to declare context dependencies or as a key for DI.
+ * @beta
+ */
+export type FASTContext<T> = ContextDecorator<T> & {
+    get(target: EventTarget): T;
+    request(target: EventTarget, callback: ContextCallback<T>, multiple?: boolean): void;
+    handle(target: EventTarget, callback: (event: ContextEvent<FASTContext<T>>) => void);
+};
+
+/**
+ * A strategy that controls how all Context.request API calls are handled.
+ * @remarks
+ * By default this is handled via Context.dispatch, which dispatched a ContextEvent.
+ * @beta
+ */
+export type FASTContextRequestStrategy = <T extends UnknownContext>(
+    target: EventTarget,
+    context: T,
+    callback: ContextCallback<ContextType<T>>,
+    multiple
+) => void;
+
+const contextEventType = "context-request";
+let requestStrategy: FASTContextRequestStrategy;
+
+/**
+ * Enables using the {@link https://github.com/webcomponents-cg/community-protocols/blob/main/proposals/context.md | W3C Community Context protocol.}
+ * @beta
+ */
+export const Context = Object.freeze({
+    /**
+     * The event type used for W3C Context Protocol requests.
+     */
+    eventType: contextEventType,
+
+    /**
+     * Creates a W3C Community Protocol-based Context object to use in requesting/providing
+     * context through the DOM.
+     * @param name - The name to use for the connext. Useful in debugging.
+     * @param initialValue - An optional initial value to use if a context handler isn't found.
+     */
+    create<T = unknown>(name: string, initialValue?: T): FASTContext<T> {
+        const Interface = function (
+            target: Constructable<Node>,
+            property: string,
+            index: number
+        ): void {
+            if (target == null || new.target !== undefined) {
+                throw new Error(`No registration for context: '${Interface.name}'`);
```

Comment:
Do we have a "philosophy" around when we should/should not throw vs. fail silently?  I've been thinking of adding errors in places (anchored region "no anchor specified").

Reply:
I'm not sure we have hard/fast rules on that, but in this case, this is a hard fail scenario. It shouldn't be possible to get the code into this state unless someone is doing something very odd and if it does get here, the decorator definitely won't be able to function.

I do have a system that was recently added to fast-element where error codes are used so that we can split out the error messages into a debug module. I'd like to propagate that through to foundation but it's a separate set of work.